### PR TITLE
Develop Compatible Triggers

### PR DESCRIPTION
**Pull request by kessler-frost**
- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation and CHANGELOG accordingly.
- [x] I have read the CONTRIBUTING document.

### CODE REVIEW

File: covalent/_dispatcher_plugins/local.py [hunk 1]
```diff
@@ -189,9 +226,63 @@ def func(*new_args, **new_kwargs):
...
+
+        stop_triggers_url = f"{triggers_server_addr}/api/triggers/stop_observe"
+
+        if isinstance(dispatch_ids, str):
+            dispatch_ids = [dispatch_ids]
+
+        r = requests.post(stop_triggers_url, json=dispatch_ids)
+        r.raise_for_status()
+
+        print("Triggers for following dispatch_ids have stopped observing:")
```
Comment (santoshkumarradha) [inline covalent/_dispatcher_plugins/local.py:286]: @kessler-frost can we remove this print statement and adding it in as log statements?

Comment (kessler-frost) [reply covalent/_dispatcher_plugins/local.py]: Removed 👍

**Review by cjao — COMMENTED** [review]
Hi @kessler-frost , could you summarize the design somewhere? Some aspects are a bit subtle -- such as the fact that the ` triggers` property of a lattice isn't persisted server-side, but is rather sent to the Trigger server right before workflow submission and then discarded from the lattice.

File: covalent/_results_manager/results_manager.py [hunk 2]
```diff
@@ -39,17 +39,19 @@
 log_stack_info = logger.log_stack_info
 
 
-def get_result(dispatch_id: str, wait: bool = False, dispatcher_addr: str = None) -> Result:
+def get_result(
+    dispatch_id: str, wait: bool = False, dispatcher_addr: str = None, status_only: bool = False
```
Comment (FyzHsn) [inline covalent/_results_manager/results_manager.py:43]: Add status_only to the doc strings

File: covalent/_workflow/lattice.py [hunk 3]
```diff
@@ -321,6 +323,7 @@ def lattice(
     deps_pip: Union[DepsPip, list] = None,
     call_before: Union[List[DepsCall], DepsCall] = [],
     call_after: Union[List[DepsCall], DepsCall] = [],
+    triggers: Union["BaseTrigger", List["BaseTrigger"]] = None,
```
Comment (FyzHsn) [inline covalent/_workflow/lattice.py:326]: Can't you use the BaseTigger object you imported in the TYPE_CHECKING import section instead of the string?

File: covalent/triggers/dir_trigger.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        depending on which ones are requested by the user.
+        Args:
+            event_names: List of event names upon which to perform a trigger action
+        """
+
+        app_log.warning("Attaching methods to dir handler")
+
+        self.n_changes = 0
+
+        def proxy_trigger(_, event):
```
Comment (FyzHsn) [inline covalent/triggers/dir_trigger.py:105]: Add unit test for this function.

File: tests/functional_tests/triggers_test.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+"""
+
+import time
+from pathlib import Path
+
+import covalent as ct
+from covalent.triggers import DirTrigger, TimeTrigger
+
+
+def test_dir_trigger():
```
Comment (FyzHsn) [inline tests/functional_tests/triggers_test.py:32]: Add one-liner doc string

File: tests/functional_tests/triggers_test.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    Path(read_file_path).unlink()
+    Path("./tr_read").rmdir()
+    Path(write_file_path).unlink()
+    Path("./tr_write").rmdir()
+
+    assert expected_sums == actual_sums
+
+
+def test_time_trigger():
```
Comment (FyzHsn) [inline tests/functional_tests/triggers_test.py:85]: Add one-liner doc string

File: tests/covalent_tests/triggers/trigger_loader_test.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    new_trigger_loader = TriggerLoader()
+
+    if not attr_is_present:
+        super_mock.assert_called_once()
+        assert new_trigger_loader != trigger_loader
+    else:
+        assert new_trigger_loader == trigger_loader
+
+
+def test_trigger_loader_get_and_set(mocker, trigger_loader):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/trigger_loader_test.py:70]: Add one-liner doc string

File: tests/covalent_tests/triggers/trigger_loader_test.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    )
+
+    assert all(
+        expected_values[i] in trigger_loader.available_triggers.values()
+        for i in range(len(expected_values))
+    )
+
+
+@pytest.mark.parametrize("attr_is_present", [True, False])
+def test_trigger_loader_new(mocker, trigger_loader, attr_is_present):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/trigger_loader_test.py:52]: Add one-liner doc string

File: tests/covalent_tests/triggers/trigger_loader_test.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+from covalent.triggers import BaseTrigger, DirTrigger, TimeTrigger
+from covalent.triggers.trigger_loader import TriggerLoader
+
+
+@pytest.fixture
+def trigger_loader():
+    return TriggerLoader()
+
+
+def test_trigger_loader_init(trigger_loader):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/trigger_loader_test.py:35]: Add one-liner doc string

File: tests/covalent_tests/triggers/trigger_loader_test.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+from unittest import mock
+
+import pytest
+
+from covalent.triggers import BaseTrigger, DirTrigger, TimeTrigger
+from covalent.triggers.trigger_loader import TriggerLoader
+
+
+@pytest.fixture
+def trigger_loader():
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/trigger_loader_test.py:31]: Add one-liner doc string

File: tests/covalent_tests/triggers/time_trigger_test.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+    if not set_stop:
+        event_is_set_mock.call_count == 2
+        sleep_mock.assert_called_with(test_time_gap)
+        trigger_mock.assert_called_once()
+    else:
+        event_is_set_mock.assert_called_once()
+
+
+def test_stop(time_trigger):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/time_trigger_test.py:67]: Add one-liner doc string

File: tests/covalent_tests/triggers/time_trigger_test.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+def is_set_side_effect(set_stop):
+    global counter
+    if set_stop or counter == 1:
+        return True
+    counter += 1
+    return False
+
+
+@pytest.mark.parametrize("set_stop", [False, True])
+def test_observe(mocker, time_trigger, set_stop):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/time_trigger_test.py:45]: Add one-liner doc string

File: tests/covalent_tests/triggers/time_trigger_test.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+counter = 0
+
+
+@pytest.fixture
+def time_trigger():
+    return TimeTrigger(None)
+
+
+def is_set_side_effect(set_stop):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/time_trigger_test.py:36]: Add one-liner doc string

File: tests/covalent_tests/triggers/time_trigger_test.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+import pytest
+
+from covalent.triggers import TimeTrigger
+
+counter = 0
+
+
+@pytest.fixture
+def time_trigger():
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/time_trigger_test.py:32]: Add one-liner doc string

File: tests/covalent_tests/triggers/dir_trigger_test.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    dir_trigger.observe()
+
+    event_loop_mock.assert_called_once()
+    event_handler_mock.assert_called_once()
+    attach_mock.assert_called_once()
+    observer_schedule_mock.assert_called_once()
+    observer_start_mock.assert_called_once()
+
+
+def test_stop(dir_trigger):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/dir_trigger_test.py:85]: Add one-liner doc string

File: tests/covalent_tests/triggers/dir_trigger_test.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    setattr_mock = mocker.patch("covalent.triggers.dir_trigger.setattr")
+
+    dir_trigger.event_handler = event_handler_mock
+    dir_trigger.attach_methods_to_handler()
+
+    assert setattr_mock.call_count == len(dir_trigger.event_names)
+
+
+@pytest.mark.parametrize("test_recursive", [True, False])
+def test_observe(mocker, dir_trigger, test_recursive):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/dir_trigger_test.py:66]: Add one-liner doc string

File: tests/covalent_tests/triggers/dir_trigger_test.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    return DirTrigger(test_dir_path, test_event_names)
+
+
+def test_event_handler_init():
+    dir_event_handler = DirEventHandler()
+    for k, v in dir_event_handler.supported_event_to_func_names.items():
+        assert v == f"on_{k}"
+
+
+def test_attach_methods_to_handler(mocker, event_to_func_names, dir_trigger):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/dir_trigger_test.py:52]: Add one-liner doc string

File: tests/covalent_tests/triggers/dir_trigger_test.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+
+@pytest.fixture
+def dir_trigger(event_to_func_names):
+    test_dir_path = "test_dir_path"
+    test_event_names = list(event_to_func_names.keys())[:4]
+    return DirTrigger(test_dir_path, test_event_names)
+
+
+def test_event_handler_init():
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/dir_trigger_test.py:46]: Add one-liner doc string

File: tests/covalent_tests/triggers/dir_trigger_test.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    return {
+        "test_a": "test_func_a",
+        "test_b": "test_func_b",
+        "test_c": "test_func_c",
+        "test_d": "test_func_d",
+        "test_e": "test_func_e",
+    }
+
+
+@pytest.fixture
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/dir_trigger_test.py:39]: Add one-liner doc string

File: tests/covalent_tests/triggers/dir_trigger_test.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+from unittest import mock
+
+import pytest
+
+from covalent.triggers.dir_trigger import DirEventHandler, DirTrigger
+
+
+@pytest.fixture
+def event_to_func_names():
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/dir_trigger_test.py:29]: Add one-liner doc string

File: tests/covalent_tests/triggers/base_test.py [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+        redispatch_id = base_trigger._do_redispatch(is_pending)
+
+        mock_redispatch.assert_called_once()
+
+    assert redispatch_id == mock_redispatch_id
+
+
+@pytest.mark.parametrize("lattice_dispatch_id", [None, "mock_id"])
+@pytest.mark.parametrize("mock_status", [None, "some_status"])
+def test_trigger(mocker, lattice_dispatch_id, mock_status):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/base_test.py:119]: Add one-liner doc string

File: tests/covalent_tests/triggers/base_test.py [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+        mock_get_status = mocker.patch("covalent.get_result", return_value=mock_status)
+        status = base_trigger._get_status()
+        mock_get_status.assert_called_once()
+
+    assert status == "mocked-status"
+
+
+@pytest.mark.parametrize("use_internal_func", [True, False])
+@pytest.mark.parametrize("is_pending", [True, False])
+def test_do_redispatch(mocker, use_internal_func, is_pending):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/base_test.py:89]: Add one-liner doc string

File: tests/covalent_tests/triggers/base_test.py [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+
+@pytest.mark.parametrize(
+    "use_internal_func, mock_status",
+    [
+        (True, JSONResponse("mock")),
+        (True, {"status": "mocked-status"}),
+        (False, {"status": "mocked-status"}),
+    ],
+)
+def test_get_status(mocker, use_internal_func, mock_status):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/base_test.py:53]: Add one-liner doc string

File: tests/covalent_tests/triggers/base_test.py [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+
+from unittest import mock
+
+import pytest
+from fastapi.responses import JSONResponse
+
+from covalent.triggers import BaseTrigger
+
+
+def test_register(mocker):
```
Comment (FyzHsn) [inline tests/covalent_tests/triggers/base_test.py:29]: Add one-liner doc string

File: tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+        else:
+            trigger_mock.observe.assert_called_once()
+
+        test_active_triggers[test_dispatch_id] = [trigger_mock]
+        assert test_active_triggers == active_triggers_mock
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("disable_triggers", [True, False])
+async def test_stop_observe(
```
Comment (FyzHsn) [inline tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py:144]: Add one-liner doc string

File: tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+    assert setattr_mock.call_count == 2
+    assert len(trigger_dict) == 2
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("disable_triggers", [True, False])
+@pytest.mark.parametrize("use_internal_funcs", [True, False])
+@pytest.mark.parametrize("observe_blocks", [True, False])
+@pytest.mark.parametrize("contains_triggers", [True, False])
+async def test_register_and_observe(
```
Comment (FyzHsn) [inline tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py:94]: Add one-liner doc string

File: tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+
+    first_threadpool = get_threadpool()
+    second_threadpool = get_threadpool()
+
+    # Check that it's the same threadpool object
+    thread_pool_mocker.assert_called_once()
+    assert first_threadpool == second_threadpool
+
+
+def test_init_trigger(mocker: mock, file_to_test: str):
```
Comment (FyzHsn) [inline tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py:60]: Add one-liner doc string

File: tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py [hunk 28]
```diff
@@ -0,0 +1,169 @@
...
+    def __init__(self, params) -> None:
+        self.parameters = params
+
+
+@pytest.fixture
+def file_to_test():
+    return "covalent_dispatcher._triggers_app.app"
+
+
+def test_get_threadpool(mocker: mock, file_to_test: str):
```
Comment (FyzHsn) [inline tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py:49]: Add one-liner doc string

File: tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+)
+
+
+class MockSigClass:
+    def __init__(self, params) -> None:
+        self.parameters = params
+
+
+@pytest.fixture
+def file_to_test():
```
Comment (FyzHsn) [inline tests/covalent_dispatcher_tests/_triggers_app/triggers_app_test.py:45]: Add one-liner doc string

**Review by FyzHsn — CHANGES_REQUESTED** [review]
Just a superficial review - still need to look at the structure of Trigger class etc.

File: doc/source/api/triggers.rst [hunk 30]
```diff
@@ -0,0 +1,12 @@
+.. _triggers:
+
+Triggers
```
Comment (FyzHsn) [inline doc/source/api/triggers.rst:3]: Something might not have been rendered properly for the triggers API doc:

```
Performs a trigger action every time_gap seconds. :type time_gap: int :param time_gap: Amount of seconds to wait before doing a trigger action
```

**Review by FyzHsn — CHANGES_REQUESTED** [review]
Went through the docs and the RTD. Left some minor comments.

File: covalent_dispatcher/_triggers_app/app.py [hunk 31]
```diff
@@ -0,0 +1,113 @@
+# Copyright 2023 Agnostiq Inc.
```
Comment (FyzHsn) [inline covalent_dispatcher/_triggers_app/app.py:1]: Add docstrings for this module.

File: covalent_dispatcher/_triggers_app/app.py [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+
+    if trigger.observe_blocks:
+        thread_pool.submit(trigger.observe)
+    else:
+        trigger.observe()
+
+    lattice_did = trigger.lattice_dispatch_id
+
+    if active_triggers.get(lattice_did):
+        active_triggers[lattice_did].append(trigger)
```
Comment (FyzHsn) [inline covalent_dispatcher/_triggers_app/app.py:93]: Seems like an important line that you'd wanna cover in the tests

**Review by FyzHsn — CHANGES_REQUESTED** [review]
This was the last of the reviews. Overall the implementation design is very clean, with great use of classes and reworking existing functionality to integrate the triggers feature. The RTD is very thorough as well. The unit tests are extremely satisfactory with solid mocking and checking that the methods were called with the correct arguments.

Once these comments are addressed, it's good to go on my end.

The only improvement to this would be the addition of functional tests for the trigger feature. However, given the size of the PR, it should be addressed later.

**Comment by kessler-frost** [comment]
> Hi @kessler-frost , could you summarize the design somewhere? Some aspects are a bit subtle -- such as the fact that the ` triggers` property of a lattice isn't persisted server-side, but is rather sent to the Trigger server right before workflow submission and then discarded from the lattice.

Hey @cjao , definitely I have the "TDD for Triggers" task on my to-do list.

File: .github/workflows/requirements.yml [hunk 33]
```diff
@@ -52,6 +52,7 @@ jobs:
           --ignore-module=urllib3.*
           --ignore-module=pkg_resources
           --ignore-file=covalent/executor/**
+          --ignore-file=covalent/triggers/**
```
[the remark this answers is not in the document]
Comment (kessler-frost) [reply .github/workflows/requirements.yml:55]: IIRC it was something trivial, like it didn't like conditionally accessing covalent_dispatcher or covalent. I'll have to recheck this though.

File: covalent/_workflow/lattice.py (hunk 3 above)
Comment (kessler-frost) [reply covalent/_workflow/lattice.py:326]: That is how type hinting works when we are conditionally importing things. So for example, in case `TYPE_CHECKING` is False, this line (326) is still run and will throw an error. Which is why `TYPE_CHECKING` is used along with the strings of the objects imported inside of it to have type checking without any errors.

File: doc/source/api/triggers.rst (hunk 30 above)
Comment (kessler-frost) [reply doc/source/api/triggers.rst:3]: Interesting, will take a look 🤔 .

File: doc/source/api/index.rst [hunk 34]
```diff
@@ -27,6 +28,7 @@ The following API documentation describes how to use Covalent.
 - :ref:`file_transfer` is used to queue remote or local file transfer operations prior or post electron execution.
 - :ref:`file_transfer_strategies` are used to perform download/upload/copy operations over various protocols.
 - :ref:`dask_executor` is used to execute electrons in a Dask cluster
+- :ref:`triggers` are used to execute a saved workflow depending on occurrence of a desired event.
```
[the remark this answers is not in the document]
Comment (kessler-frost) [reply doc/source/api/index.rst:31]: I think yours is better, `are used to execute a workflow triggered by a specific event type.`. I don't like "undispatched" because we do call `ct.dispatch` on the workflow.

File: tests/covalent_dispatcher_tests/_cli/service_test.py [hunk 35]
```diff
@@ -223,19 +245,9 @@ def test_graceful_shutdown_stopped_server(mocker):
...
-        (True, False, None),
-        (False, False, None),
-        (False, True, None),
-        (True, True, "112233"),
-        (True, False, "112233"),
-        (False, False, "112233"),
-        (False, True, "112233"),
-    ],
-)
+@pytest.mark.parametrize("is_migration_pending", [True, False])
```
[the remark this answers is not in the document]
Comment (kessler-frost) [reply tests/covalent_dispatcher_tests/_cli/service_test.py:248]: Yeah 😅 , pytest will automatically create a matrix based on the parametrizations.

File: tests/functional_tests/triggers_test.py (hunk 5 above)
Comment (kessler-frost) [reply tests/functional_tests/triggers_test.py:32]: @FyzHsn this is the file containing the functional tests for the triggers feature.

Comment (FyzHsn) [reply tests/functional_tests/triggers_test.py:32]: ah I see 😅

File: tests/functional_tests/triggers_test.py [hunk 36]
```diff
@@ -0,0 +1,114 @@
...
+
+
+def test_dir_trigger():
+    Path("./tr_read").mkdir(exist_ok=True)
+    Path("./tr_write").mkdir(exist_ok=True)
+
+    read_file_path = str(Path("./tr_read/trigger_read.txt").resolve())
+    write_file_path = str(Path("./tr_write/trigger_write.txt").resolve())
+
+    print(read_file_path)
```
Comment (kessler-frost) [inline tests/functional_tests/triggers_test.py:39]: Need to remove this.

Comment (kessler-frost) [reply tests/functional_tests/triggers_test.py]: Removed.

**Comment by FyzHsn** [comment]
Address the RTD API rendering issue in a separate PR

**Review by FyzHsn — APPROVED** [review]

**Review by FyzHsn — DISMISSED** [review]

File: covalent_ui/webapp/src/components/dispatches/ResultListing.js [hunk 37]
```diff
@@ -907,11 +907,15 @@ const ResultListing = () => {
                             />
                           </TableCell>
                         ) : (
-                          <TableCell>{secondsToHms(result.runTime)}</TableCell>
+                          <TableCell>{result.runTime === null ? '-' : secondsToHms(result.runTime)}</TableCell>
```
Comment (AlejandroEsquivel) [inline covalent_ui/webapp/src/components/dispatches/ResultListing.js:910]: Minor comment that we should check for the falsy value of `result.runTime` to ensure `secondsToHms` doesn't receive falsy values

```javascript
{ !result.runTime ? '-' : secondsToHms(result.runTime)}
```

This can be addressed along with the exception handling changes on a follow up PR

File: covalent_dispatcher/_triggers_app/app.py [hunk 38]
```diff
@@ -0,0 +1,133 @@
...
+    else:
+        active_triggers[lattice_did] = [trigger]
+
+    app_log.debug(f"Started trigger with id: {lattice_did}")
+
+
+@router.post("/triggers/stop_observe")
+async def stop_observe(request: Request):
+    """
```
Comment (AlejandroEsquivel) [inline covalent_dispatcher/_triggers_app/app.py:118]: @kessler-frost it might be good wrap try except blocks around all of the route handler to catch any uncaught exceptions, and then re-raise a `HTTPException` if there's any errors so that we can get a proper HTTP JSON response (FastAPI ultimately catches `HTTPException` and we can have a middleware to render error responses according to our schema)

I will be moving that into a separate issue however so shouldn't block this 👍🏼

File: covalent_ui/webapp/src/components/dispatches/ResultListing.js (hunk 37 above)
Comment (AlejandroEsquivel) [reply covalent_ui/webapp/src/components/dispatches/ResultListing.js]: I just made the change here so we can disregard above 👍🏼
